File: generators/app/templates/_PROJECT_NAME_-srv/handler.js
const debug = require('debug')('<%= projectName %>-srv:function');
<% if(authentication || hana){ -%>
const xsenv = require('@sap/xsenv');
const services = xsenv.getServices({
<% if(authentication){ -%>
  uaa: { label: 'xsuaa' }
<% } -%>
<% if(hana){ -%>
<% if(authentication){ -%>
  ,
<% } -%>
  hana: { label: 'hana' }
<% } -%>
});

<% } -%>
<% if(authentication){ -%>
const util = require('util');
const xssec = require('@sap/xssec');
const createSecurityContext = util.promisify(xssec.createSecurityContext);

<% } -%>
<% if(hana){ -%>
const hana = require('@sap/hana-client');
services.hana.sslValidateCertificate = true;
services.hana.ssltruststore = services.hana.certificate;
const hanaConn = hana.createConnection();

async function execDB(sql, isPrepared=false, params) {
  try {
    await hanaConn.connect(services.hana);
  } catch (err) {
    debug('execDB connect', err.message, err.stack);
    results = err.message;
  }
  try {
    await hanaConn.exec('SET SCHEMA ' + services.hana.schema);
    if (isPrepared) {
      let hanaStmt = await hanaConn.prepare(sql);
      results = hanaStmt.exec(params);
    } else {
      results = await hanaConn.exec(sql);
    }
  } catch (err) {
    debug('execDB exec', err.message, err.stack);
    results = err.message;
  }
  try {
    await hanaConn.disconnect();
  } catch (err) {
    debug('execDB disconnect', err.message, err.stack);
    results = err.message;
  }
  return results;
}

<% } -%>
<% if(apiGraph || apiDest){ -%>
const httpClient = require('@sap-cloud-sdk/http-client');
<% } -%>
<% if(apiS4HC || apiGraph || apiDest){ -%>
const { retrieveJwt } = require('@sap-cloud-sdk/connectivity');
<% } -%>
<% if(apiS4HC){ -%>
const { desc } = require('@sap-cloud-sdk/odata-v2');
const { salesOrderService } = require('@sap/cloud-sdk-vdm-sales-order-service');
const { salesOrderApi, salesOrderItemApi } = salesOrderService();
async function getSalesOrders(req) {
  return salesOrderApi.requestBuilder()
    .getAll()
    .filter(salesOrderApi.schema.TOTAL_NET_AMOUNT.greaterThan(2000))
    .top(3)
    .orderBy(desc(salesOrderApi.schema.LAST_CHANGE_DATE_TIME))
    .select(
      salesOrderApi.schema.SALES_ORDER,
      salesOrderApi.schema.LAST_CHANGE_DATE_TIME,
      salesOrderApi.schema.INCOTERMS_LOCATION_1,
      salesOrderApi.schema.TOTAL_NET_AMOUNT,
      salesOrderApi.schema.TO_ITEM.select(salesOrderItemApi.schema.MATERIAL, salesOrderItemApi.schema.NET_AMOUNT)
    )
    .execute({
      destinationName: '<%= projectName %>-s4hc-api'
<% if(authentication){ -%>
      ,
      jwt: retrieveJwt(req)
<% } -%>
    });
}

<% } -%>
module.exports = {
  main: async function (event, context) {
<% if(subscription){ -%>
    if (typeof event.data.eventType === 'string') {
      switch(event.data.eventType) {
        case 'defaultEvent':
          debug('Event', event.data.eventType, event.data.value);
          break;
<% if(subscription){ -%>
        case 'salesBoost':
          debug('Event', event.data.eventType, event.data.id, event.data.amount);
          await execDB(`UPDATE "<%= projectName %>.db::sales" SET "amount" = "amount" + ? WHERE "id" = ?`,true,[event.data.amount, event.data.id]);
          break;
<% } -%>
        default:
          debug('Event unknown', event.data.eventType);
      }
      event.extensions.response.sendStatus(200);
      return;
    }

<% } -%>
    let req = event.extensions.request;

<% if(authentication){ -%>
    let securityContext;
    if (typeof req.headers.authorization === 'string' && req.headers.authorization.split(' ').length > 1 && req.headers.authorization.split(' ')[0].toLowerCase() === 'bearer') {
      try {
        securityContext = await createSecurityContext(req.headers.authorization.split(' ')[1], services.uaa);
      } catch (err) {
        debug('Create Security Context', err.message);
        event.extensions.response.sendStatus(401);
        return;
      }
    } else {
      debug('Create Security Context', 'Invalid Headers - Missing Access Token');
      event.extensions.response.sendStatus(401);
      return;
    }

<% } -%>
    switch(req.path) {
      case '/srv/':
<% if(authorization){ -%>
        if (securityContext.checkScope('$XSAPPNAME.User')) {
<% } -%>
          results = '<%= projectName %>';
<% if(authorization){ -%>
        } else {
          event.extensions.response.sendStatus(403);
          return;
        }
<% } -%>
        break;

<% if(authentication){ -%>
      case '/srv/user':
<% if(authorization){ -%>
        if (securityContext.checkScope('$XSAPPNAME.User')) {
<% } -%>
          try {
            let user = {};
            user.logonName = securityContext.getLogonName();
            user.givenName = securityContext.getGivenName();
            user.familyName = securityContext.getFamilyName();
            user.email = securityContext.getEmail();
<% if(authorization){ -%>
            user.scopes = {};
            user.scopes.User = securityContext.checkScope('$XSAPPNAME.User');
            user.scopes.Admin = securityContext.checkScope('$XSAPPNAME.Admin');
<% } -%>
            results = user;
          } catch (err) {
            debug('/srv/user', err.message, err.stack);
            results = err.message;
          }
<% if(authorization){ -%>
        } else {
          event.extensions.response.sendStatus(403);
          return;
        }
<% } -%>
        break;

<% } -%>
<% if(apiS4HC){ -%>
      case '/srv/salesorders':
<% if(authorization){ -%>
        if (securityContext.checkScope('$XSAPPNAME.User')) {
<% } -%>
          results = await getSalesOrders(req);
<% if(authorization){ -%>
        } else {
          event.extensions.response.sendStatus(403);
          return;
        }
<% } -%>
        break;

<% } -%>
<% if(apiGraph){ -%>
      case '/srv/graph':
<% if(authorization){ -%>
        if (securityContext.checkScope('$XSAPPNAME.User')) {
<% } -%>
          try {
            let res1 = await httpClient.executeHttpRequest(
              {
                destinationName: '<%= projectName %>-graph-api'
<% if(authentication){ -%>
                ,
                jwt: retrieveJwt(req)
<% } -%>
              },
              {
                method: 'GET',
                url: req.query.path || '<%= apiGraphId %>'
              }
            );
            let location = req.query.location || '';
            if (location !== '') {
              let res2 = await httpClient.executeHttpRequest(
                {
                  destinationName: '<%= projectName %>-graph-api'
<% if(authentication){ -%>
                  ,
                  jwt: retrieveJwt(req)
<% } -%>
                },
                {
                  method: 'PATCH',
                  url: req.query.path || '<%= apiGraphId %>',
                  headers: {
                    "If-Match": res1.headers.etag
                  },
                  data: {
                    "IncotermsTransferLocation": location,
                    "IncotermsLocation1": location
                  }
                },
                {
                  fetchCsrfToken: false
                }
              );
              results = res2.data;
            } else {
              results = res1.data;
            };
          } catch (err) {
            debug('/srv/graph', err.message, err.stack);
            results = err.message;
          }
<% if(authorization){ -%>
        } else {
          event.extensions.response.sendStatus(403);
          return;
        }
<% } -%>
        break;

<% } -%>
<% if(apiDest){ -%>
      case '/srv/dest':
<% if(authorization){ -%>
        if (securityContext.checkScope('$XSAPPNAME.User')) {
<% } -%>
          try {
            let res1 = await httpClient.executeHttpRequest(
              {
                destinationName: req.query.destination || '<%= projectName %>-nw'
<% if(authentication){ -%>
                ,
                jwt: retrieveJwt(req)
<% } -%>
              },
              {
                method: 'GET',
                url: req.query.path || ''
              }
            );
            results = res1.data;
          } catch (err) {
            debug('/srv/dest', err.message, err.stack);
            results = err.message;
          }
<% if(authorization){ -%>
        } else {
          event.extensions.response.sendStatus(403);
          return;
        }
<% } -%>
        break;

<% } -%>
<% if(hana){ -%>
      case '/srv/sales':
<% if(authorization){ -%>
        if (securityContext.checkScope('$XSAPPNAME.User')) {
<% } -%>
          results = await execDB(`SELECT * FROM "<%= projectName %>.db::sales"`);
<% if(authorization){ -%>
        } else {
          event.extensions.response.sendStatus(403);
          return;
        }
<% } -%>
        break;

      case '/srv/topSales':
<% if(authorization){ -%>
        if (securityContext.checkScope('$XSAPPNAME.User')) {
<% } -%>
          let amount = req.query.amount ?? 0;
          results = await execDB(`CALL "<%= projectName %>.db::SP_TopSales"(?,?)`,true, [amount]);
<% if(authorization){ -%>
        } else {
          event.extensions.response.sendStatus(403);
          return;
        }
<% } -%>
        break;

      case '/srv/session':
<% if(authorization){ -%>
        if (securityContext.checkScope('$XSAPPNAME.Admin')) {
<% } -%>
          results = await execDB(`SELECT * FROM M_SESSION_CONTEXT`);
<% if(authorization){ -%>
        } else {
          event.extensions.response.sendStatus(403);
          return;
        }
<% } -%>
        break;

      case '/srv/db':
<% if(authorization){ -%>
        if (securityContext.checkScope('$XSAPPNAME.Admin')) {
<% } -%>
          results = await execDB(`SELECT SYSTEM_ID, DATABASE_NAME, HOST, VERSION, USAGE FROM M_DATABASE`);
<% if(authorization){ -%>
        } else {
          event.extensions.response.sendStatus(403);
          return;
        }
<% } -%>
        break;

      case '/srv/connections':
<% if(authorization){ -%>
        if (securityContext.checkScope('$XSAPPNAME.Admin')) {
<% } -%>
          results = await execDB(`SELECT TOP 10 USER_NAME, CLIENT_IP, CLIENT_HOST, START_TIME FROM M_CONNECTIONS WHERE OWN='TRUE' ORDER BY START_TIME DESC`);
<% if(authorization){ -%>
        } else {
          event.extensions.response.sendStatus(403);
          return;
        }
<% } -%>
        break;

<% } -%>
      default:
        event.extensions.response.sendStatus(400);
        return;
    }

    return results;
  }
}

File: generators/app/templates/_PROJECT_NAME_-srv/handler.py
import bottle

<% if(authentication || hana || apiDest){ -%>
import os
from os import path

<% } -%>
<% if(authentication){ -%>
from sap import xssec
UAA_ROOT = os.environ.get("SERVICE_BINDING_ROOT") + "/<%= projectName %>-uaa"
with open(path.join(UAA_ROOT, "clientid"), "r") as f:
  uaa_clientid = f.read()
with open(path.join(UAA_ROOT, "clientsecret"), "r") as f:
  uaa_clientsecret = f.read()
with open(path.join(UAA_ROOT, "url"), "r") as f:
  uaa_url = f.read()
with open(path.join(UAA_ROOT, "verificationkey"), "r") as f:
  uaa_verificationkey = f.read()
with open(path.join(UAA_ROOT, "xsappname"), "r") as f:
  uaa_xsappname = f.read()
uaa_service = {"clientid": uaa_clientid, "clientsecret": uaa_clientsecret, "url": uaa_url, "verificationkey": uaa_verificationkey, "xsappname": uaa_xsappname}

<% } -%>
<% if(apiDest){ -%>
import requests
DEST_ROOT = os.environ.get("SERVICE_BINDING_ROOT") + "/<%= projectName %>-dest"
with open(path.join(DEST_ROOT, "clientid"), "r") as f:
  dest_clientid = f.read()
with open(path.join(DEST_ROOT, "clientsecret"), "r") as f:
  dest_clientsecret = f.read()
with open(path.join(DEST_ROOT, "uri"), "r") as f:
  dest_uri = f.read()
with open(path.join(DEST_ROOT, "url"), "r") as f:
  dest_url = f.read()
def api_dest(destination, path=""):
  token = requests.post(dest_url + "/oauth/token?grant_type=client_credentials", auth=(dest_clientid,dest_clientsecret)).json()["access_token"]
  url = requests.get(dest_uri + "/destination-configuration/v1/destinations/" + destination, headers={"Authorization": "Bearer " + token}).json()["destinationConfiguration"]["URL"]
  results = requests.get(url + "/" + path, headers={"Accept": "Application/json"}).json()
  return results

<% } -%>
<% if(hana){ -%>
import json
from hdbcli import dbapi
HDI_ROOT = os.environ.get("SERVICE_BINDING_ROOT") + "/<%= projectName %>-hdi"
with open(path.join(HDI_ROOT, "host"), "r") as f:
  hana_host = f.read()
with open(path.join(HDI_ROOT, "port"), "r") as f:
  hana_port = f.read()
with open(path.join(HDI_ROOT, "schema"), "r") as f:
  hana_schema = f.read()
with open(path.join(HDI_ROOT, "user"), "r") as f:
  hana_user = f.read()
with open(path.join(HDI_ROOT, "password"), "r") as f:
  hana_password = f.read()

def exec_db(sql, is_procedure=False, is_prepared=False, params=""):
  try:
    conn = dbapi.connect(address=hana_host, port=int(hana_port), encrypt=True, user=hana_user, password=hana_password, currentSchema=hana_schema)
  except dbapi.Error as err:
    results = format(err)
  else:
    with conn.cursor() as cursor:
      try:
        if is_procedure:
          cursor.callproc(sql,(params,0))
        elif is_prepared:
          cursor.execute(sql,params)
        else:
          cursor.execute(sql)
      except dbapi.Error as err:
        results = format(err)
      else:
        if is_procedure or sql.lower().split()[0] == 'select':
          rows = cursor.fetchall()
          # convert to JSON
          results = []
          for row in rows:
            this_row = {}
            for index, column in enumerate(row):
              this_row[cursor.description[index][0]] = column
            results.append(this_row)
          results = json.dumps(results)
        else:
          results = ''
        cursor.close()
    try:
      conn.close()
    except dbapi.Error as err:
      results = format(err)
  return results

<% } -%>
def main(event, context):
<% if(subscription){ -%>
  if type(event["ce-type"]) is str:
    if event["ce-type"] == "sap.kyma.custom.<%= projectName %>.default.event.v1":
      print("Event", event["ce-type"], event["data"]["value"])
      return bottle.HTTPResponse(status=200, headers={"content-type": "text/plain"})
<% if(hana){ -%>
    elif event["ce-type"] == "sap.kyma.custom.<%= projectName %>.sales.boost.v1":
      print("Event", event["ce-type"], event["data"]["id"], event["data"]["amount"])
      exec_db('UPDATE "<%= projectName %>.db::sales" SET "amount" = "amount" + :amount WHERE "id" = :id', False, True, {"id": event["data"]["id"], "amount": event["data"]["amount"]})
      return bottle.HTTPResponse(status=200, headers={"content-type": "text/plain"})
<% } -%>
    else:
      print("Event unknown", event["ce-type"])
      return bottle.HTTPResponse(status=200, headers={"content-type": "text/plain"})

<% } -%>
<% if(authentication){ -%>
  token = event["extensions"]["request"].headers.get("Authorization")
  if token:
    token = token.split(" ")[1]
  else:
    return bottle.HTTPResponse(status=401, headers={"content-type": "text/plain"})
  try:
    security_context = xssec.create_security_context(token, uaa_service)
  except:
    return bottle.HTTPResponse(status=401, headers={"content-type": "text/plain"})

<% } -%>
  url_path = event["extensions"]["request"].path
  if url_path == "/srv/":
<% if(authorization){ -%>
    if security_context.check_scope("$XSAPPNAME.User"):
<% } -%>
      results = '<%= projectName %>';
<% if(authorization){ -%>
    else:
      return bottle.HTTPResponse(status=403, headers={"content-type": "text/plain"})
<% } -%>

<% if(authentication){ -%>
  elif url_path == "/srv/user":
<% if(authorization){ -%>
    if security_context.check_scope("$XSAPPNAME.User"):
<% } -%>
      results = {"logonName": security_context.get_logon_name(), "givenName": security_context.get_given_name(), "familyName": security_context.get_family_name(), "email": security_context.get_email()<% if(authorization){ -%>, "scopeUser": security_context.check_scope("$XSAPPNAME.User"), "scopeAdmin": security_context.check_scope("$XSAPPNAME.Admin")<% } -%>}
<% if(authorization){ -%>
    else:
      return bottle.HTTPResponse(status=403, headers={"content-type": "text/plain"})
<% } -%>

<% } -%>
<% if(apiDest){ -%>
  elif url_path == "/srv/dest":
<% if(authorization){ -%>
    if security_context.check_scope("$XSAPPNAME.User"):
<% } -%>
      destination = event["extensions"]["request"].query.destination
      if not destination:
        destination = "<%= projectName %>-nw"
      path = event["extensions"]["request"].query.path
      if not path:
        path = ""
      results = api_dest(destination, path)
<% if(authorization){ -%>
    else:
      return bottle.HTTPResponse(status=403, headers={"content-type": "text/plain"})
<% } -%>

<% } -%>
<% if(hana){ -%>
  elif url_path == "/srv/sales":
<% if(authorization){ -%>
    if security_context.check_scope("$XSAPPNAME.User"):
<% } -%>
      results = exec_db('SELECT * FROM "<%= projectName %>.db::sales"')
<% if(authorization){ -%>
    else:
      return bottle.HTTPResponse(status=403, headers={"content-type": "text/plain"})
<% } -%>

  elif url_path == "/srv/topSales":
<% if(authorization){ -%>
    if security_context.check_scope("$XSAPPNAME.User"):
<% } -%>
      amount = event["extensions"]["request"].query.amount
      if not amount:
        amount = 0
      results = exec_db('"<%= projectName %>.db::SP_TopSales"', True, False, amount)
<% if(authorization){ -%>
    else:
      return bottle.HTTPResponse(status=403, headers={"content-type": "text/plain"})
<% } -%>

  elif url_path == "/srv/session":
<% if(authorization){ -%>
    if security_context.check_scope("$XSAPPNAME.Admin"):
<% } -%>
      results = exec_db("SELECT * FROM M_SESSION_CONTEXT")
<% if(authorization){ -%>
    else:
      return bottle.HTTPResponse(status=403, headers={"content-type": "text/plain"})
<% } -%>

  elif url_path == "/srv/db":
<% if(authorization){ -%>
    if security_context.check_scope("$XSAPPNAME.Admin"):
<% } -%>
      results = exec_db("SELECT SYSTEM_ID, DATABASE_NAME, HOST, VERSION, USAGE FROM M_DATABASE")
<% if(authorization){ -%>
    else:
      return bottle.HTTPResponse(status=403, headers={"content-type": "text/plain"})
<% } -%>

  elif url_path == "/srv/connections":
<% if(authorization){ -%>
    if security_context.check_scope("$XSAPPNAME.Admin"):
<% } -%>
      results = exec_db("SELECT TOP 10 USER_NAME, CLIENT_IP, CLIENT_HOST, TO_NVARCHAR(START_TIME) AS START_TIME FROM M_CONNECTIONS WHERE OWN='TRUE' ORDER BY START_TIME DESC")
<% if(authorization){ -%>
    else:
      return bottle.HTTPResponse(status=403, headers={"content-type": "text/plain"})
<% } -%>

<% } -%>
  else:
    return bottle.HTTPResponse(status=400, headers={"content-type": "text/plain"})

  return results